tracer/src/Datadog.Trace/Util/ThrowHelper.cs

```diff
@@ -16,56 +18,67 @@ internal class ThrowHelper
         [MethodImpl(MethodImplOptions.NoInlining)]
         [DebuggerHidden]
         [DoesNotReturn]
-        internal static void ThrowArgumentNullException(string paramName) => throw new ArgumentNullException(paramName);
+        internal static void ThrowArgumentNullException(string? paramName) => throw new ArgumentNullException(paramName);
+
+        /// <summary>Throws an <see cref="ArgumentNullException"/> if <paramref name="argument"/> is null.</summary>
+        /// <param name="argument">The reference type argument to validate as non-null.</param>
+        /// <param name="paramName">The name of the parameter with which <paramref name="argument"/> corresponds.</param>
+        public static void ThrowArgumentNullExceptionIfNull<T>(T? argument, [CallerArgumentExpression("argument")] string? paramName = null)
```

Comment:
dotnet implemented this with `object?`, but that can cause boxing, so I used `T?`. WDYT?

Reply:
I like the idea of using generics to avoid boxing, but in this case a value type cannot be null, so we are going to call this method only for reference types, and for that `object` is just fine. 

I see this solution on RangeExceptions or something similar that can work for both `Object` and `ValueType`

Reply:
I will run some benchmarks to compare what happens when `T` is a reference type vs a value type. We're not sure if the boxing happens anyway during the null comparison. Thanks!